google/cloud/bigtable/_row_merger.py

```diff
@@ -0,0 +1,627 @@
+    ) -> None:
+        """
+        Helper function to push items from an async generator into a cache
+        """
+        async for item in input_generator:
+            await cache.put(item)
+
+    async def retryable_merge_rows(
+        self,
+        gapic_fn,
+        cache_size,
+        per_row_timeout,
+        per_request_timeout,
+        revise_on_retry,
+        row_limit,
+    ) -> AsyncGenerator[Row | RequestStats, None]:
+        """
+        Retryable wrapper for merge_rows. This function is called each time
+        a retry is attempted.
+
+        Some fresh state is created on each retry:
+          - grpc network stream
+          - cache for the stream
+          - state machine to hold merge chunks received from stream
+        Some state is shared between retries:
+          - last_seen_row_key is used to ensure that
+            duplicate rows are not emitted
+          - request is stored and (optionally) modified on each retry
+        """
+        if revise_on_retry and self.last_seen_row_key is not None:
+            # if this is a retry, try to trim down the request to avoid ones we've already processed
+            self.request["rows"] = _RowMerger._revise_request_rowset(
+                row_set=self.request.get("rows", None),
+                last_seen_row_key=self.last_seen_row_key,
+            )
+            # revise next request's row limit based on number emitted
+            if row_limit:
+                new_limit = row_limit - self.emit_count
+                if new_limit <= 0:
+                    return
+                else:
+                    self.request["rows_limit"] = new_limit
+        new_gapic_stream = await gapic_fn(
+            self.request,
+            timeout=per_request_timeout,
+        )
+        cache: asyncio.Queue[Row | RequestStats] = asyncio.Queue(maxsize=cache_size)
+        state_machine = _StateMachine()
+        try:
+            stream_task = asyncio.create_task(
+                _RowMerger._generator_to_cache(
+                    cache,
+                    _RowMerger.merge_row_response_stream(
+                        new_gapic_stream, state_machine
+                    ),
+                )
+            )
+            get_from_cache_task = asyncio.create_task(cache.get())
+            # sleep to allow other tasks to run
+            await asyncio.sleep(0)
+            # read from state machine and push into cache
+            # when finished, stream will be done, cache will be empty, but get_from_cache_task will still be waiting
+            while (
+                not stream_task.done()
+                or not cache.empty()
+                or get_from_cache_task.done()
+            ):
+                if get_from_cache_task.done():
+                    new_item = get_from_cache_task.result()
+                    # don't yield rows that have already been emitted
+                    if isinstance(new_item, RequestStats):
+                        yield new_item
```

Comment:
what happens with RequestStats on a retry?

Reply:
I plan on fleshing out the request_stats feature more in a separate PR, but currently the generator will hold the most recently returned RequestStats object.

Do we need to combine the data from multiple RequestStats if available? 

My understanding was that the stats objects would be the last ones returned in a stream though, so getting multiple should be pretty rare, right?